### PR TITLE
chore(observability): extend component validation framework for more flexible test case building

### PR DESCRIPTION
**Pull request by neuronull**
- Added a way to construct the test case log events from the YAML definition on a per-field basis (before we were always injecting into the `message` field). This was necessary in order to construct events in a way that the Splunk HEC protocol expects. It will likely continue to be useful.
- Some components requires a specific HTTP headers. As such, the external resource configuration for Http resources, was extended to allow specifying custom headers to add to events.
- Removed the assumption that all components would be handling events in a `Vec` . This is significant because some components only operate on a per-event basis. And the `EstimatedJsonEncodedSizeOf` calculation on a `Vec` , includes 2 bytes per bracket character... thus the testing framework needs to handle this. The solution I created for this is to allow a value for this metric that either matches within a Vec or without. We could alternatively try encoding this information into the `ComponentTestCaseConfiguration`.

### CODE REVIEW

File: src/components/validation/resources/event.rs [hunk 1]
```diff
@@ -36,17 +38,26 @@ pub enum RawTestEvent {
...
 
 impl EventData {
     /// Converts this event data into an `Event`.
     pub fn into_event(self) -> Event {
         match self {
             Self::Log(message) => Event::Log(LogEvent::from_bytes_legacy(&message.into())),
+            Self::LogBuilder(data) => {
+                let mut log_event = LogEvent::default();
+                for (k, v) in data {
+                    log_event.parse_path_and_insert(k, v).unwrap();
```
Comment (kapilsinha) [inline src/components/validation/resources/event.rs:57]: Is it safe to do a raw `unwrap()` here?

File: src/components/validation/resources/http.rs [hunk 2]
```diff
@@ -219,16 +229,23 @@ fn spawn_input_http_client(
...
             let mut buffer = BytesMut::new();
             encode_test_event(&mut encoder, &mut buffer, event);
 
-            let request = Request::builder()
+            let mut request_builder = Request::builder()
                 .uri(request_uri.clone())
-                .method(request_method.clone())
+                .method(request_method.clone());
+
+            for header in &headers {
```
Comment (kapilsinha) [inline src/components/validation/resources/http.rs:244]: Cleaner to destructure? i.e.
`for (key, value) in &headers`

File: src/components/validation/runner/mod.rs [hunk 3]
```diff
@@ -205,6 +205,14 @@ impl Runner {
 
         let test_cases = load_component_test_cases(&self.test_case_data_path)?;
         for test_case in test_cases {
+            println!("");
```
Comment (kapilsinha) [inline src/components/validation/runner/mod.rs:208]: Remove?

File: src/components/validation/validators/component_spec/mod.rs [hunk 4]
```diff
@@ -263,7 +283,15 @@ fn compare_actual_to_expected(
...
     info!("{}: expected {}, actual {}.", metric_type, expected, actual,);
 
-    if actual != expected {
+    if actual != expected &&
+        // This is a bit messy. The issue is that EstimatedJsonSizeOf can be called by a component
+        // on an event array, or on a single event. And we have no way of knowing which that is.
+        // By default the input driver for the framework is not assuming it is an array, so we
+        // check here if it matches what the array scenario would be, which is to add the size of
+        // the brackets, for each event.
+        (metric_type != &ComponentMetricType::EventsReceivedBytes
```
Comment (kapilsinha) [inline src/components/validation/validators/component_spec/mod.rs:292]: Is there test coverage for this case?

File: src/components/validation/resources/event.rs (hunk 1 above)
Comment (neuronull) [reply src/components/validation/resources/event.rs:57]: this is all code in a test framework so its much safer than production code. But I'll change this to an `expect` and give a more helpful error message.

File: src/components/validation/runner/mod.rs (hunk 3 above)
Comment (neuronull) [reply src/components/validation/runner/mod.rs:208]: this is intentional to add line breaks between the test cases.

File: src/components/validation/validators/component_spec/mod.rs (hunk 4 above)
Comment (neuronull) [reply src/components/validation/validators/component_spec/mod.rs:292]: This is a test framework code, so this is the test coverage...

This branch contains changes to the framework that are utilized in downstream branches. it was done this way to separate the framework changes from usage to make it more reviewable.

Comment (kapilsinha) [reply src/components/validation/validators/component_spec/mod.rs:292]: Ha my bad, thanks for clarifying. Pretty sure I was looking at a component validation PR just a few days ago, but had forgotten that context

**Review by tobz — APPROVED** [review]
The bit about having to try an additional calculation to match the single event vs vec of events stuff isn't wonderful, but it's not the worst thing in the codebase. 🤷🏻

**Pull request merged**